platform-api/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/stack/StackService.java

```diff
@@ -0,0 +1,422 @@
+import static java.lang.String.format;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+import static javax.ws.rs.core.MediaType.TEXT_PLAIN;
+import static javax.ws.rs.core.MediaType.MULTIPART_FORM_DATA;
+import static javax.ws.rs.core.Response.Status.CREATED;
+import static org.eclipse.che.api.workspace.server.Constants.LINK_REL_CREATE_STACK;
+import static org.eclipse.che.api.workspace.server.Constants.LINK_REL_GET_STACKS_BY_CREATOR;
+import static org.eclipse.che.api.workspace.server.Constants.LINK_REL_GET_STACK_BY_ID;
+import static org.eclipse.che.api.workspace.server.Constants.LINK_REL_REMOVE_STACK;
+import static org.eclipse.che.api.workspace.server.Constants.LINK_REL_SEARCH_STACKS;
+import static org.eclipse.che.api.workspace.server.Constants.LINK_REL_UPDATE_STACK;
+import static org.eclipse.che.api.workspace.server.Constants.LINK_REL_UPLOAD_ICON;
+import static org.eclipse.che.api.workspace.server.Constants.LINK_REL_GET_ICON;
+import static org.eclipse.che.api.workspace.server.Constants.LINK_REL_DELETE_ICON;
+import static org.eclipse.che.api.workspace.server.DtoConverter.asDto;
+
+/**
+ * Defines Stack REST API
+ *
+ * @author Alexander Andrinko
+ */
+@Api(value = "/stack", description = "Stack REST API")
+@Path("/stack")
+public class StackService extends Service {
+
+    private final StackDao           stackDao;
+    private final PermissionsChecker permissionChecker;
+
+    @Inject
+    public StackService(StackDao stackDao, PermissionsChecker permissionChecker) {
+        this.stackDao = stackDao;
+        this.permissionChecker = permissionChecker;
+    }
+
+    @POST
+    @Consumes(APPLICATION_JSON)
+    @Produces(APPLICATION_JSON)
+    @GenerateLink(rel = LINK_REL_CREATE_STACK)
+    @RolesAllowed({"user", "system/admin", "system/manager"})
+    @ApiOperation(value = "Create a new stack",
+                  notes = "This operation can be performed only by authorized user",
+                  response = StackDtoDescriptor.class)
+    @ApiResponses({@ApiResponse(code = 201, message = "The stack successfully created"),
+                   @ApiResponse(code = 400, message = "Missed required parameters, parameters are not valid"),
+                   @ApiResponse(code = 403, message = "The user does not have access to create a new stack"),
+                   @ApiResponse(code = 409, message = "Conflict error occurred during the stack creation" +
+                                                      "(e.g. The stack with such name already exists)"),
+                   @ApiResponse(code = 500, message = "Internal server error occurred")})
+    public Response createStack(@ApiParam("The new stack") StackDto stackDto) throws ApiException {
+        requireNonNull(stackDto, "Stack required");
+        requireNonNullAndNonEmpty(stackDto.getName(), "Stack name required");
+        if (stackDto.getSource() == null && stackDto.getWorkspaceConfig() == null) {
+            throw new BadRequestException("Stack source required");
```

Comment:
If stack source is not null but workspace config is, then message still will be `Stack source required` which will confuse user